engine/src/main/java/io/camunda/zeebe/engine/state/EventApplier.java

```diff
@@ -29,6 +29,34 @@ public interface EventApplier {
    * @param intent the intent of the event
    * @param recordValue the value of the event
    * @param recordVersion the record version of the event
+   * @throws NoSuchEventApplier if no event applier is found for the given intent and record
+   *     version. The event is not applied and it is up to the caller to decide what to do.
    */
-  void applyState(long key, Intent intent, RecordValue recordValue, final int recordVersion);
+  void applyState(long key, Intent intent, RecordValue recordValue, final int recordVersion)
+      throws NoSuchEventApplier;
+
+  /** Thrown when no event applier is found for a given intent and record version. */
+  abstract sealed class NoSuchEventApplier extends RuntimeException {
+    public NoSuchEventApplier(final String message) {
+      super(message);
+    }
+
+    public static final class NoApplierForIntent extends NoSuchEventApplier {
+      public NoApplierForIntent(final Intent intent) {
+        super(
+            String.format(
+                "Expected to find an event applier for intent '%s', but none was found.", intent));
+      }
+    }
+
+    public static final class NoApplierForVersion extends NoSuchEventApplier {
+      public NoApplierForVersion(
+          final Intent intent, final int recordVersion, final int latestVersion) {
+        super(
+            String.format(
+                "Expected to find an event applier for intent '%s' and version '%d', but '%s' is the latest supported version.",
+                intent, recordVersion, latestVersion));
+      }
+    }
```

Comment:
❓ SM users would see this error message. Is it clear enough to figure out how to recover? I think so right, because of https://github.com/camunda/zeebe/pull/16160#issuecomment-1938712611

Reply:
Yeah I think it's okay. Usually this would only show up during a rolling update (if at all) and would resolve itself as the rolling update proceeds.

Reply:
💭 It could also occur on downgrading.